File: Classes/@BIP/BIP.m
classdef BIP
% BIP  Binary Interaction Parameters container.
%
%   bip = BIP(n) creates an n-component BIP object with all parameter
%   matrices initialised to zero. Set individual matrices after
%   construction to provide non-zero interaction parameters.
%
%   Properties (all n-by-n matrices unless noted):
%     EOScons, EOStdep                  - EOS kij parameters
%     NRTLcons, NRTLtdep, NRTLtdep2    - NRTL: A + B*T + C*T^2
%     NRTLtdepm1, NRTLtdeplog          - NRTL: D/T + E*ln(T) terms
%     NRTLalfa                          - NRTL non-randomness parameter
%     Wilsoncons, Wilsontdep            - Wilson parameters
%     UNIQUACcons, UNIQUACtdep         - UNIQUAC interaction parameters
%     UNIQUACR  (1-by-n)               - UNIQUAC volume parameters
%     UNIQUACQ  (1-by-n)               - UNIQUAC surface parameters

%{
Copyright (c) 2012, 2013, Ali Akbar Eftekhari
All rights reserved.

Redistribution and use in source and binary forms, with or
without modification, are permitted provided that the following
conditions are met:

    *   Redistributions of source code must retain the above copyright notice,
        this list of conditions and the following disclaimer.
    *   Redistributions in binary form must reproduce the above
        copyright notice, this list of conditions and the following
        disclaimer in the documentation and/or other materials provided
        with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
%}

    properties
        EOScons
        EOStdep
        NRTLcons
        NRTLtdep
        NRTLtdep2
        NRTLtdepm1
        NRTLtdeplog
        NRTLalfa
        Wilsoncons
        Wilsontdep
        UNIQUACcons
        UNIQUACtdep
        UNIQUACR
        UNIQUACQ
    end

    methods
        function obj = BIP(n)
            obj.EOScons     = zeros(n);
            obj.EOStdep     = zeros(n);
            obj.NRTLcons    = zeros(n);
            obj.NRTLtdep    = zeros(n);
            obj.NRTLtdep2   = zeros(n);
            obj.NRTLtdepm1  = zeros(n);
            obj.NRTLtdeplog = zeros(n);
            obj.NRTLalfa    = zeros(n);
            obj.Wilsoncons  = zeros(n);
            obj.Wilsontdep  = zeros(n);
            obj.UNIQUACcons = zeros(n);
            obj.UNIQUACtdep = zeros(n);
            obj.UNIQUACR    = zeros(1, n);
            obj.UNIQUACQ    = zeros(1, n);
        end
    end
end

File: pvt/bip.py
"""BIP — Binary Interaction Parameters container.

Stores all matrices for binary interaction parameters used by equations of state
and activity coefficient models. All fields are zero-initialised for an
n-component mixture.

Matrices are n×n symmetric numpy arrays. UNIQUAC R and Q are 1D arrays of length n.
"""

from dataclasses import dataclass, field

import numpy as np


@dataclass
class BIP:
    """Binary Interaction Parameters for an n-component mixture.

    All interaction matrices are initialised to zero. Set non-zero values
    by assigning to individual fields after construction.

    EOS kij(T) = EOScons + EOStdep * T
    NRTL A(T) = NRTLcons + NRTLtdep*T + NRTLtdep2*T² + NRTLtdepm1/T + NRTLtdeplog*ln(T)
    Wilson A(T) = Wilsoncons + Wilsontdep*T
    UNIQUAC A(T) = UNIQUACcons + UNIQUACtdep*T

    Attributes:
        EOScons: Constant part of EOS kij.
        EOStdep: Temperature-dependent part of EOS kij.
        NRTLcons: Constant part of NRTL A_ij.
        NRTLtdep: B coefficient for NRTL (T term).
        NRTLtdep2: C coefficient for NRTL (T² term).
        NRTLtdepm1: D coefficient for NRTL (1/T term).
        NRTLtdeplog: E coefficient for NRTL (ln T term).
        NRTLalfa: Non-randomness parameter α_ij.
        Wilsoncons: Constant part of Wilson A_ij.
        Wilsontdep: Temperature-dependent part of Wilson A_ij.
        UNIQUACcons: Constant part of UNIQUAC A_ij.
        UNIQUACtdep: Temperature-dependent part of UNIQUAC A_ij.
        UNIQUACR: UNIQUAC volume parameter r_i (1D array).
        UNIQUACQ: UNIQUAC surface area parameter q_i (1D array).
    """

    EOScons: np.ndarray
    EOStdep: np.ndarray
    NRTLcons: np.ndarray
    NRTLtdep: np.ndarray
    NRTLtdep2: np.ndarray
    NRTLtdepm1: np.ndarray
    NRTLtdeplog: np.ndarray
    NRTLalfa: np.ndarray
    Wilsoncons: np.ndarray
    Wilsontdep: np.ndarray
    UNIQUACcons: np.ndarray
    UNIQUACtdep: np.ndarray
    UNIQUACR: np.ndarray
    UNIQUACQ: np.ndarray

    def __init__(self, n: int):
        """Initialise all BIP matrices to zero for n components.

        Args:
            n: Number of components in the mixture.
        """
        self.EOScons = np.zeros((n, n))
        self.EOStdep = np.zeros((n, n))
        self.NRTLcons = np.zeros((n, n))
        self.NRTLtdep = np.zeros((n, n))
        self.NRTLtdep2 = np.zeros((n, n))
        self.NRTLtdepm1 = np.zeros((n, n))
        self.NRTLtdeplog = np.zeros((n, n))
        self.NRTLalfa = np.zeros((n, n))
        self.Wilsoncons = np.zeros((n, n))
        self.Wilsontdep = np.zeros((n, n))
        self.UNIQUACcons = np.zeros((n, n))
        self.UNIQUACtdep = np.zeros((n, n))
        self.UNIQUACR = np.zeros(n)
        self.UNIQUACQ = np.zeros(n)
